sigmastate/src/main/scala/sigmastate/Values.scala

```diff
@@ -1522,4 +1523,121 @@ object Values {
+          ContractTemplate(treeVersion, name, description, constTypes, constValues, parameters, expressionTree, null)
+        )
+      )
+
+    @inline final def isRightParsed: Boolean = expressionTree.isRight
+
+    private[sigmastate] var _bytes: Array[Byte] = templateBytes
+
+    /** Serialized bytes of this template. */
+    final def bytes: Array[Byte] = {
+      if (_bytes == null) {
+        _bytes = ContractTemplateSerializer.DefaultSerializer.serializeContractTemplate(this)
+      }
+      _bytes
+    }
+
+    def applyTemplate(paramValues: IndexedSeq[Constant[SType]]): ErgoTree = {
+      require(paramValues.length == parameters.length,
+        s"Expected ${parameters.length} number of parameters, got ${paramValues.length}")
+      require(isRightParsed, "expressionTree must be parsable to be able to apply template")
+
+      val nConsts = constTypes.size
+      val constants = safeNewArray[Constant[SType]](nConsts)
+      val paramIndices = this.parameters.map(p => p.placeholder).toSet
+      val positionToParamIndex = this.parameters.indices.map(i => (i -> this.parameters(i).placeholder)).toMap
+
+//      cfor(0)(_ < nConsts, _ + 1) { i =>
+//        if ()
+//      }
+
+      val constValues = this.constValues match {
+        case Some(values) => values
+        case None => safeNewArray[Option[SType#WrappedType]](constTypes.size)
+      }
+
+      cfor(0)(_ < paramValues.length, _ + 1) { i =>
+        val pos = parameters(i).placeholder
+//        constValues(pos) = paramValues(i)
+      }
+
+      // TODO: Check that value and param types match.
+
+      ErgoTree(
+        ErgoTree.ConstantSegregationHeader,
+        Array[Constant[SType]](),
+        this.expressionTree.right.get
+      )
+    }
+
+    /** The default equality of case class is overridden to exclude `complexity`. */
+    override def canEqual(that: Any): Boolean = that.isInstanceOf[ContractTemplate]
+
+    override def hashCode(): Int = Objects.hash(treeVersion, name, description, constTypes, constValues, parameters, expressionTree)
+
+    override def equals(obj: Any): Boolean = (this eq obj.asInstanceOf[AnyRef]) ||
+      ((obj.asInstanceOf[AnyRef] != null) && (obj match {
+        case other: ContractTemplate =>
+          other.treeVersion == treeVersion && other.name == name && other.description == description && other.constTypes == constTypes && other.constValues == constValues && other.parameters == parameters && other.expressionTree == expressionTree
+        case _ => false
+      }))
+  }
+
+  object ContractTemplate {
+    def apply(name: String,
+              description: String,
+              constTypes: IndexedSeq[SType],
+              constValues: Option[IndexedSeq[Option[SType#WrappedType]]],
+              parameters: IndexedSeq[Parameter],
+              expressionTree: SigmaPropValue): ContractTemplate = {
+      new ContractTemplate(None, name, description, constTypes, constValues, parameters, Right(expressionTree))
+    }
+  }
```

Comment:
Please rebase on top of https://github.com/ScorexFoundation/sigmastate-interpreter/compare/sdk-separate-interpreters
and move all the new code into `sdk` module.

Reply:
@aslesarenko Need some help on this. I need some of the code from `sc` module in the `sdk` module. Example is `JsonCodecs`, `DataJsonEncoder`, `SigmaParser` etc.

I tried adding the dependency for `sc` in the build.sbt but it fails as `sc` is not a `crossProject`. Can you please advise on how to proceed here?

1. I can make `sc` a `crossProject`
2. I can move the needed logic outside to some other module - `corelib` or `interpreter` which are already a dependency for `sdk`.

Thanks!

Reply:
sdk module shouldn't depend on sc.
I think you can move JsonCodecs, DataJsonEncoder to sdk module.
But, it is better we keep SigmaParser out of sdk.
Do you really need it for eip5?

Reply:
I need SigmaParser to decode SType. 

Basically this:
``` scala
implicit val sTypeDecoder: Decoder[SType] = Decoder.instance({ implicit cursor =>
      fromTry(Try.apply(SigmaParser.parseType(cursor.value.asString.get)))
    })
```

Also note that, `DataJsonEncoder` depends on `SigmaParser` for the same -- to decode the SType. See [here](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/f6db2bbf3f4ddcc07d18d23bc01f557c23290264/sc/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala#L266-L283).

I thought about copying the `SigmaParser.parseType` functionality into `sdk` module but just copying [these 9-10 lines](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/f6db2bbf3f4ddcc07d18d23bc01f557c23290264/sc/src/main/scala/sigmastate/lang/SigmaParser.scala#L104-L112) isn't enough as it depends on the definition of `Type` which is present in `Types.scala` and the dependencies go on and on.

May be we should move the `SigmaParser` functionality into a shared module?

Reply:
Fixed by #864 and #867